File: testing/Android/app/src/main/java/com/styleconverter/test/style/borders/sides/BorderSideExtractor.kt
package com.styleconverter.test.style.borders.sides

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import com.styleconverter.test.style.PropertyRegistry
import com.styleconverter.test.style.core.types.ValueExtractors
import kotlinx.serialization.json.JsonElement

/**
 * Extracts border side configurations from IR properties.
 * Handles both shorthand (border-width) and individual side properties.
 */
object BorderSideExtractor {

    init {
        // Claim every physical + logical border-side property so the
        // legacy dispatch knows we own them. Categories:
        //   * shorthand (BorderWidth/Color/Style)
        //   * physical sides (Top/Right/Bottom/Left × Width/Color/Style)
        //   * logical sides (BlockStart/BlockEnd/InlineStart/InlineEnd × …)
        // CSS spec: https://drafts.csswg.org/css-backgrounds-3/#borders
        //           https://drafts.csswg.org/css-logical/#border-properties
        PropertyRegistry.migrated(
            "BorderWidth", "BorderColor", "BorderStyle",
            "BorderTopWidth", "BorderRightWidth", "BorderBottomWidth", "BorderLeftWidth",
            "BorderTopColor", "BorderRightColor", "BorderBottomColor", "BorderLeftColor",
            "BorderTopStyle", "BorderRightStyle", "BorderBottomStyle", "BorderLeftStyle",
            "BorderBlockStartWidth", "BorderBlockEndWidth",
            "BorderInlineStartWidth", "BorderInlineEndWidth",
            "BorderBlockStartColor", "BorderBlockEndColor",
            "BorderInlineStartColor", "BorderInlineEndColor",
            "BorderBlockStartStyle", "BorderBlockEndStyle",
            "BorderInlineStartStyle", "BorderInlineEndStyle",
            owner = "borders/sides"
        )
    }

    /**
     * Extract border configuration from a list of property type/data pairs.
     *
     * @param properties List of pairs where first is the property type (e.g., "BorderTopWidth")
     *                   and second is the JSON data for that property.
     * @return AllBordersConfig with extracted values for all sides.
     */
    fun extractBorderConfig(properties: List<Pair<String, JsonElement?>>): AllBordersConfig {
        var top = BorderSideConfig()
        var end = BorderSideConfig()
        var bottom = BorderSideConfig()
        var start = BorderSideConfig()

        // Track shorthand values to apply to sides without specific values
        var sharedWidth: Dp? = null
        var sharedColor: Color? = null
        var sharedStyle: ValueExtractors.LineStyle? = null

        properties.forEach { (type, data) ->
            when (type) {
                // Shorthand (all sides)
                "BorderWidth" -> sharedWidth = ValueExtractors.extractBorderWidth(data)
                "BorderColor" -> sharedColor = ValueExtractors.extractColor(data)
                "BorderStyle" -> sharedStyle = ValueExtractors.extractLineStyle(data)

                // Top side (physical and logical)
                "BorderTopWidth", "BorderBlockStartWidth" -> top = top.copy(width = ValueExtractors.extractBorderWidth(data))
                "BorderTopColor", "BorderBlockStartColor" -> top = top.copy(color = ValueExtractors.extractColor(data))
                "BorderTopStyle", "BorderBlockStartStyle" -> top = top.copy(style = ValueExtractors.extractLineStyle(data))

                // Right/End side (physical and logical)
                "BorderRightWidth", "BorderInlineEndWidth" -> end = end.copy(width = ValueExtractors.extractBorderWidth(data))
                "BorderRightColor", "BorderInlineEndColor" -> end = end.copy(color = ValueExtractors.extractColor(data))
                "BorderRightStyle", "BorderInlineEndStyle" -> end = end.copy(style = ValueExtractors.extractLineStyle(data))

                // Bottom side (physical and logical)
                "BorderBottomWidth", "BorderBlockEndWidth" -> bottom = bottom.copy(width = ValueExtractors.extractBorderWidth(data))
                "BorderBottomColor", "BorderBlockEndColor" -> bottom = bottom.copy(color = ValueExtractors.extractColor(data))
                "BorderBottomStyle", "BorderBlockEndStyle" -> bottom = bottom.copy(style = ValueExtractors.extractLineStyle(data))

                // Left/Start side (physical and logical)
                "BorderLeftWidth", "BorderInlineStartWidth" -> start = start.copy(width = ValueExtractors.extractBorderWidth(data))
                "BorderLeftColor", "BorderInlineStartColor" -> start = start.copy(color = ValueExtractors.extractColor(data))
                "BorderLeftStyle", "BorderInlineStartStyle" -> start = start.copy(style = ValueExtractors.extractLineStyle(data))
            }
        }

        // Apply shorthand values to sides that don't have specific values
        if (sharedWidth != null || sharedColor != null || sharedStyle != null) {
            top = top.copy(
                width = top.width ?: sharedWidth,
                color = top.color ?: sharedColor,
                style = top.style ?: sharedStyle
            )
            end = end.copy(
                width = end.width ?: sharedWidth,
                color = end.color ?: sharedColor,
                style = end.style ?: sharedStyle
            )
            bottom = bottom.copy(
                width = bottom.width ?: sharedWidth,
                color = bottom.color ?: sharedColor,
                style = bottom.style ?: sharedStyle
            )
            start = start.copy(
                width = start.width ?: sharedWidth,
                color = start.color ?: sharedColor,
                style = start.style ?: sharedStyle
            )
        }

        return AllBordersConfig(top, end, bottom, start)
    }
}

File: runtimes/swiftui/Sources/StyleConverterRuntime/StyleEngine/borders/sides/BorderSideExtractor.swift
//
//  BorderSideExtractor.swift
//  StyleEngine/borders/sides — Phase 5.
//
//  Walks a property list once and fills an AllBordersConfig. Handles all
//  24 `border-*` side properties: 4 physical sides × {width, color, style}
//  plus 4 logical sides × {width, color, style}. Shorthand longhands
//  (`BorderWidth`, `BorderColor`, `BorderStyle`) fan out to every side
//  that didn't set an explicit value — matches the parser registry in
//  `src/main/kotlin/app/parsing/css/properties/shorthands/borders/sides/`.
//

// Foundation + CoreGraphics (via SwiftUI) for Color/CGFloat.
import SwiftUI

// Single static entry point to keep the call site clean — mirrors
// ColorExtractor / PaddingExtractor.
enum BorderSideExtractor {

    // List of property-type strings this extractor owns. The registry
    // consults this list (via `PropertyRegistry.migrated`) so the legacy
    // StyleBuilder switch skips them.
    static let propertyNames: [String] = [
        // Shorthand longhands — CSS 2.1 §8.5 per-family fallthroughs.
        "BorderWidth", "BorderColor", "BorderStyle",
        // Physical sides.
        "BorderTopWidth", "BorderTopColor", "BorderTopStyle",
        "BorderRightWidth", "BorderRightColor", "BorderRightStyle",
        "BorderBottomWidth", "BorderBottomColor", "BorderBottomStyle",
        "BorderLeftWidth", "BorderLeftColor", "BorderLeftStyle",
        // Logical sides — CSS Logical Properties Level 1.
        "BorderBlockStartWidth", "BorderBlockStartColor", "BorderBlockStartStyle",
        "BorderBlockEndWidth", "BorderBlockEndColor", "BorderBlockEndStyle",
        "BorderInlineStartWidth", "BorderInlineStartColor", "BorderInlineStartStyle",
        "BorderInlineEndWidth", "BorderInlineEndColor", "BorderInlineEndStyle",
    ]

    // Returns nil when no border-* property appears in the list — lets the
    // applier short-circuit without allocating.
    static func extract(from properties: [IRProperty]) -> AllBordersConfig? {
        // Per-side accumulators. Logical names map into the same four
        // buckets since we assume LTR (fixtures confirm this).
        var top = BorderSideConfig()
        var end = BorderSideConfig()
        var bottom = BorderSideConfig()
        var start = BorderSideConfig()
        // Shorthand-derived fallbacks. Applied after the explicit pass so
        // per-side overrides win — mirrors the Android extractor.
        var sharedWidth: CGFloat? = nil
        var sharedColor: Color? = nil
        var sharedStyle: BorderStyleValue? = nil
        // Touch flag — returns nil when no border-* entry was seen so the
        // applier skips dispatch entirely.
        var touched = false

        for prop in properties {
            switch prop.type {
            // ── Shorthand longhands (apply-to-all-sides fallback) ──
            case "BorderWidth":
                // Width keywords (`thin|medium|thick`) are pre-resolved to
                // `{ "px": N }` by the CSS parser before reaching iOS.
                sharedWidth = ValueExtractors.extractPx(prop.data); touched = true
            case "BorderColor":
                sharedColor = ValueExtractors.extractColor(prop.data); touched = true
            case "BorderStyle":
                sharedStyle = extractBorderStyle(prop.data); touched = true

            // ── Top (physical) + Block-start (logical, LTR = top) ──
            case "BorderTopWidth", "BorderBlockStartWidth":
                top.width = ValueExtractors.extractPx(prop.data); touched = true
            case "BorderTopColor", "BorderBlockStartColor":
                top.color = ValueExtractors.extractColor(prop.data); touched = true
            case "BorderTopStyle", "BorderBlockStartStyle":
                top.style = extractBorderStyle(prop.data); touched = true

            // ── Right (physical) + Inline-end (logical, LTR = right) ──
            case "BorderRightWidth", "BorderInlineEndWidth":
                end.width = ValueExtractors.extractPx(prop.data); touched = true
            case "BorderRightColor", "BorderInlineEndColor":
                end.color = ValueExtractors.extractColor(prop.data); touched = true
            case "BorderRightStyle", "BorderInlineEndStyle":
                end.style = extractBorderStyle(prop.data); touched = true

            // ── Bottom (physical) + Block-end (logical, LTR = bottom) ──
            case "BorderBottomWidth", "BorderBlockEndWidth":
                bottom.width = ValueExtractors.extractPx(prop.data); touched = true
            case "BorderBottomColor", "BorderBlockEndColor":
                bottom.color = ValueExtractors.extractColor(prop.data); touched = true
            case "BorderBottomStyle", "BorderBlockEndStyle":
                bottom.style = extractBorderStyle(prop.data); touched = true

            // ── Left (physical) + Inline-start (logical, LTR = left) ──
            case "BorderLeftWidth", "BorderInlineStartWidth":
                start.width = ValueExtractors.extractPx(prop.data); touched = true
            case "BorderLeftColor", "BorderInlineStartColor":
                start.color = ValueExtractors.extractColor(prop.data); touched = true
            case "BorderLeftStyle", "BorderInlineStartStyle":
                start.style = extractBorderStyle(prop.data); touched = true

            default:
                // Not ours.
                break
            }
        }

        if !touched { return nil }

        // Shorthand fan-out — fill in any side that didn't set an explicit
        // value of its own. Mirrors the Android extractor's semantics so
        // the cross-platform coverage comparison is byte-for-byte aligned.
        if sharedWidth != nil || sharedColor != nil || sharedStyle != nil {
            if top.width == nil    { top.width    = sharedWidth }
            if top.color == nil    { top.color    = sharedColor }
            if top.style == nil    { top.style    = sharedStyle }
            if end.width == nil    { end.width    = sharedWidth }
            if end.color == nil    { end.color    = sharedColor }
            if end.style == nil    { end.style    = sharedStyle }
            if bottom.width == nil { bottom.width = sharedWidth }
            if bottom.color == nil { bottom.color = sharedColor }
            if bottom.style == nil { bottom.style = sharedStyle }
            if start.width == nil  { start.width  = sharedWidth }
            if start.color == nil  { start.color  = sharedColor }
            if start.style == nil  { start.style  = sharedStyle }
        }

        return AllBordersConfig(top: top, end: end, bottom: bottom, start: start)
    }
}
